### PR TITLE
Functionality to convert between R and Python

### PR DESCRIPTION
**Pull request by jorainer**
This PR based on the work of @michaelwitting and:
- expands the R-Python-R conversion functions to allow specifying which variables to copy and how to map the variable/metadata names between R and Python.
- add and expand documentation.
- use the `basilisk` package to ensure the *correct* Python environment is available to the package (will be automatically installed and bundled within the package - i.e. is independent on any system Python packages).
- expand the vignette.
- add unit tests for all functions.

### CODE REVIEW

**Comment by jorainer** [comment]
Maybe also @chufz and @hechth could have a look at this to see if it is OK?

**Comment by jorainer** [comment]
This is related to issue #8

File: DESCRIPTION [hunk 1]
```diff
@@ -20,22 +18,31 @@ Authors@R:
...
 VignetteBuilder: knitr
-BugReports: https://github.com/RforMassSpectrometry/MetaboAnnotation/issues
-URL: https://github.com/RforMassSpectrometry/MetaboAnnotation
+BugReports: https://github.com/RforMassSpectrometry/SpectriPy/issues
+URL: https://github.com/RforMassSpectrometry/SpectriPy
 biocViews: Infrastructure, Metabolomics, MassSpectrometry
-Roxygen: list(markdown=TRUE)
+Encoding: UTF-8
+StagedInstall: no
+SystemRequirements: python (>= 2.7)
```
Comment (hechth) [inline DESCRIPTION:46]: I'd stick with python larger 3.5 or so - not sure how well things work in python2

File: R/basilisk.R [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#' @importFrom basilisk BasiliskEnvironment
+matchms_env <- basilisk::BasiliskEnvironment(
+    envname = "matchms_env", pkgname = "SpectriPy",
+    packages = c("matchms==0.14.0"),
+    channels = c("bioconda", "conda-forge")
+)
```
Comment (hechth) [inline R/basilisk.R:6]: This is pretty nice!

File: R/conversion.R [hunk 3]
```diff
@@ -1,94 +1,169 @@
...
+#' @export
+rspec_to_pyspec <- function(x, mapping = spectraVariableMapping(),
+                            reference = import("matchms"),
+                            BPPARAM = SerialParam(), .check = TRUE) {
+    if (.check && !is(x, "Spectra"))
+        stop("'x' should be a Spectra object.")
+    plist <- spectrapply(x, .single_rspec_to_pyspec, spectraVariables = mapping,
+                         BPPARAM = BPPARAM)
+    r_to_py(unname(plist))
 }
```
Comment (hechth) [inline R/conversion.R:103]: The `reference` parameter is actually not used

File: R/conversion.R [hunk 4]
```diff
@@ -1,94 +1,169 @@
...
+    if (!is(x, "python.builtin.list"))
+        stop("'x' is expected to be a Python list.")
+    x <- py_to_r(x)
+    if (.check && !all(vapply(x, function(z)
+        is(z, "matchms.Spectrum.Spectrum"), logical(1))))
+        stop("'x' is expected to be a Python list of matchms Spectrum objects.")
+    spectra_list <- bplapply(x, .single_pyspec_to_rspec,
+                             spectraVariables = mapping, BPPARAM = BPPARAM)
+    do.call(concatenateSpectra, spectra_list)
+}
```
Comment (hechth) [inline R/conversion.R:122]: Same here, `reference` parameter is not used -> does `import("matchms")` by itself have any implications (initialize global variables) if it is called?

File: R/conversion.R [hunk 5]
```diff
@@ -98,28 +173,55 @@ convertPySpectrumToRSpectra <- function(x,
...
+        spd <- DataFrame(rlist[lengths(rlist) > 0])
+        if (!nrow(spd))
+            spd <- DataFrame(msLevel = NA_integer_)
+    } else
+        spd <- DataFrame(msLevel = NA_integer_)
+    spd$mz <- NumericList(as.numeric(x$peaks$mz), compress = FALSE)
+    spd$intensity <- NumericList(as.numeric(x$peaks$intensities),
+                                 compress = FALSE)
+    Spectra(spd)
+}
```
Comment (hechth) [inline R/conversion.R:211]: Same here - `reference` is not reference explicitly.

File: tests/testthat/test_conversion.R [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    expect_equal(intensity(res), intensity(sps[2L]))
+    expect_equal(rtime(res), rtime(sps[2L]))
+    expect_equal(msLevel(res), msLevel(sps[2L]))
+
+    p <- SpectriPy:::.single_rspec_to_pyspec(sps[3L])
+    res <- SpectriPy:::.single_pyspec_to_rspec(p, vars)
+    expect_equal(mz(res), mz(sps[3L]))
+    expect_equal(intensity(res), intensity(sps[3L]))
+    expect_equal(rtime(res), rtime(sps[3L]))
+    expect_equal(msLevel(res), msLevel(sps[3L]))
```
Comment (hechth) [inline tests/testthat/test_conversion.R:140]: This could be solved nicely using the parameterized test cases from the `patrick` R package I think.

**Review by hechth — COMMENTED** [review]
Parameterized tests would maybe make the testing cleaner, also I'm unsure about the `reference` parameter - if it is really not needed it shouldn't be there, if it has to be there (even without being used) it should maybe be stated why.

File: R/conversion.R (hunk 3 above)
Comment (michaelwitting) [reply R/conversion.R:103]: True, `reference` is not used here, but in `.single_rspec_to_pyspec`. It contains the reference to the Python Spectrum object, needed for construction. The question is if we can set a global variable that contains this reference during package loading?

File: R/basilisk.R (hunk 2 above)
Comment (jorainer) [reply R/basilisk.R:6]: yep - the `basilisk` people really did a great work to ensure reproducibility. Will make it much easier to the end user :)

File: R/conversion.R (hunk 3 above)
Comment (jorainer) [reply R/conversion.R:103]: At first I also wanted to drop it - but then I realized:

- this allows also to use `import("matchms", convert = FALSE)` to disable automatic conversion of python to R and vice versa, maybe not wrong to disable to speed up things?
- Also, in a loop, you can define this `reference` once, and then in the loop re-use it. Speeds up things.

I would thus keep it (it's anyway for the *advanced* user).

File: R/conversion.R (hunk 4 above)
Comment (jorainer) [reply R/conversion.R:122]: Now that you say it - I think I forgot to pass this `reference` down to the `.single_pyspec_to_rspec` function! I will fix that.

Comment (michaelwitting) [reply R/conversion.R:122]: We just need to reference in `reference` to the correct environment that was set up by `basilisk`, I guess the name needs to be changed to `matchms_env` then, right? so `reference = import("matchms_env")`

File: tests/testthat/test_conversion.R (hunk 6 above)
Comment (jorainer) [reply tests/testthat/test_conversion.R:140]: never used the `patrick` package thus far - I'll have a look at it.

File: R/conversion.R (hunk 5 above)
Comment (jorainer) [reply R/conversion.R]: I removed it - did not break any unit tests, so *should* be fine.

File: DESCRIPTION (hunk 1 above)
Comment (hechth) [reply DESCRIPTION]: I think 3.5 or 3.7 are reasonable choices

File: R/basilisk.R (hunk 2 above)
Comment (hechth) [reply R/basilisk.R:6]: I don't think there is a more recent one at the moment.

**Comment by jorainer** [comment]
I've addressed all your points @hechth - looking forward to your review.

**Review by hechth — APPROVED** [review]

File: R/conversion.R (hunk 3 above)
Comment (hechth) [reply R/conversion.R]: I think having a function that populates a global variable might be the best way?

I don't think that passing the reference to the imported matchms is a good thing - this implies you could use different versions of matchms in a single session etc., which is not very clean.

But we can figure out how to do this properly in another issue.

**Comment by hechth** [comment]
This should also close #10, right?

**Comment by jorainer** [comment]
Thanks for the review @hechth ! I will also try to set up proper github action tests in the main branch.

**Comment by hechth** [comment]
Do you have experience with the CI for R packages? And do we want to integrate the package into conda?

**Comment by jorainer** [comment]
Yes, for our R-packages we use github actions that check the package on Windows, macOS and Linux. Integration into conda? I have no idea, never done that :) if you think that might be good we should do it, but how about all the dependencies then? are they already in conda?

**Comment by hechth** [comment]
I can check the state of the dependencies for conda.

**Pull request merged**